### ISSUE TITLE
Normal variadic call interpreted as named-argument variadic call

### ISSUE THREAD

**Issue by gmazzap** [vimeo/psalm#5829]
Take the following example:

https://psalm.dev/r/5b8b039dd4

I'm using no named arguments, but Psalm fails with `NamedArgumentNotAllowed` (because of my `@no-named-arguments`).

Note that I use  `@no-named-arguments` to be able to declare `Test::$strings` as `list<string>`. Without excluding named arguments Psalm will (rightfully) assume it is `array<array-key, string>`

**Comment by orklah** [comment]
https://3v4l.org/NrAlf

Here, your property will be assigned to something else than a list.

You have to propagate `@no-named-arguments` to make psalm checks this:
https://psalm.dev/r/455431ab70

**Comment by gmazzap** [comment]
The problem is that I don't want to care about named parameters. If someone uses named properties with my code, it is not guaranteed to work, and I'm fine with that.

I've set `allowNamedArgumentCalls` to false, and would happy to remove _all_ `@no-named-arguments`.

But if I don't use `@no-named-arguments` then Psalm refuse to set the variadic parameter as a `list`.

What's the way to tell Psalm _"please check my code just like named parameters are not a thing"_?

**Comment by orklah** [comment]
Unfortunately, `allowNamedArgumentCalls` has no impact on the list vs array issue. It's due to how Psalm works with config files. For a library, it doesn't matters that you disabled named arguments in your own config when the customer has its own config file which allow them. Hence the phpdoc at method level

I guess a different config could be created for projects that are not libraries and that would apply project wide but this doesn't exists yet.

**Comment by orklah** [comment]
I didn't thought about code in vendor calling yours, but from the POV of an author of a piece of code that will end up in someone else vendor. From this POV, the project specific psalm config is useless to warn your users they should not use named arguments. Hence the phpdoc tag.

Your POV also makes sense and it was my own thought before I understood what was going on (and I documented it in https://github.com/vimeo/psalm/pull/5654#issue-619686506).

I personally think it could be a worthwhile addition. I also think it could be added by a plugin here, but I'm not sure which hook could be used. If you want to try, the condition that need to change seem to be this one:
https://github.com/vimeo/psalm/blob/4f9067f5c88679d859eeb0c6929ae7c8c395f041/src/Psalm/Internal/Analyzer/FunctionLikeAnalyzer.php#L1018

I'd personally try the `AfterCodebasePopulatedInterface` but I'm not totally confident on that.

**Comment by gmazzap** [comment]
> I didn't thought about code in vendor calling yours, but from the POV of an author of a piece of code that will end up in someone else vendor. From this POV, the project specific psalm config is useless to warn your users they should not use named arguments on your methods. Hence the phpdoc tag.

Yes, I understand, it's the same case I exposed from two different perspective.

From _my_ perspective, if they call my methods with named arguments, I don't care: it's their problem.

From _their_ perspective, if they call my methods with named arguments they would want to be warned, and to do that Psalm needs annotations on the method.

So the issue is IMO:

favor _me_, not forcing me to pollute my codebase with `@no-named-arguments` when I have no interest in supporting named parameters

or favor _them_, making sure they get notified when using named parameters on a methods that does not support them?

Of course, I want to favor _me_ :)

> I personally think it could be a worthwhile addition in Psalm

Great. So there's a chance someone will work on it?

> I also think it could be added by a plugin, but I'm not sure which hook could be used.

Good to know. And thank you very much for pointing me to that condition. Honestly I'm not very familiar with developing plugins for Psalm, but will surely explore as soon as I'll have some time (which, unfortunately, I don't foresee to happen soon).
